Proxy to handle dataset groups for QgsMdalProvider

vcloarec opened the pull request:
## Description
This PR allows `QgsMdalProvider` to manage dataset groups, including not valid or not compatible mesh.
To do that, a proxy is integrated to the provider. This proxy has to help the provider to return dataset data from an index. Indexes are persistent, they are not removed when dataset groups are invalid or not compatible, and provider return empty dataset groups when their indexes are used. This proxy allow objects to access the dataset groups by index even if the position of the dataset groups changes for the provider.

**New classes implemented :**

_QgsMdalDatasetIndex :_ 
Instance of this class represents an index and is associated to a datatset group loaded by the provider. Its role is to return a reference to a dataset group known by the provider.

_QgsMdalProviderDatasetProxy :_ 
Handle the QgsMeshDataProviderDataSetIndex to help the provider to retrieve the dataset groups.

The proxy is populated when the provider is created and when dataset groups are added. After, when provider is reloaded, the table present in the proxy is updated to keep acces to all the dataset group.
This approach needs to modify the` QgsMeshLayer::writeXml()` and `QgsMeshLayer::readXml()` methods, in order to save the proxy data and to consider it when loading.


## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

vcloarec committed:
Mesh provider proxy impemented

New classes implemented :

QgsMdalDatasetIndex :
Instance of this class represents an index and is associated to a datatset group loaded by the provider. Its role is to return a reference to a dataset group known by the provider.

QgsMdalProviderDatasetProxy :
Handle the QgsMeshDataProviderDataSetIndex to help the provider to retrieve the dataset groups.

writeXml and readXml methods of QgsMeshLayer have been updated to take account of this proxy. This change permits to reload all the data if the dataset files have been corrupted before the opening of the project.

vcloarec commented:
The first attempt didn't consider the mesh calculator, so unit test fail !
I tried to fix it, everything seems to be OK.
Now the calculated mesh are reloaded when re-opening project, I don't know if it is the desired behavior... If not, this behavior can be changed.

PeterPetrik reviewed (commented):
- can you please rebase the branch to one or two commits with sensible message explaining what is achieved in the commit? 
- new methods in the classes in the public API (`core/`) must he `\since QGIS 3.10 or QGIS 3.8` doxygen mark to know in which version were introduced
- please remove all commented out code
- all comments are in 3rd person (reloads, puts, gets...)
- is this PR finished or is is work in progress. If work in progress please add "WIP" to the PR name
- what is the motivation of this change? removal of the dataset groups?

Need to read the PR more closely, but as a a general question, have you consider (instead of creation of the new proxy class) changing `QgsMeshDatasetIndex` to  have "extranal/QGIS" index and "internal/MDAL/provider" index and "valid/invalid/broken" state?

vcloarec committed:
Adjustments to consider mesh calculator

Modification in previous commits doesn't consider mesh calculator.

The proxy can now handle dataset groups coming from calculation.
Correction

vcloarec commented:
Sorry, for the problems in the code, I have to make good habits and improve the git use...

**Motivation of this change :**
Discussion about the reload of dataset groups has given me the idea of this proxy and I wanted to implement it before move on.
If, the previous PR does the job for reloading the dataset group, I find not very satisfying the way the invalid dataset groups are handled, and not very clean the "mix" of renderer this invalid datasets are created, even if repairing the file permits to return in a good situation.
With this PR, even it is more complicated, I find this way cleaner and when a dataset group is invalid or incompatible, this is announced in the mesh layer properties  instead of the dataset group doesn't appear.

Furthermore, opening a project with a invalid dataset create a little issue : the invalid dataset wasn't append to the uris list, so the dataset couldn't be reloaded. The proxy doesn't directly fix this issue, but the need to save the table contained in the proxy forces to fix it.
As I have discover the mesh calculator when unit test failed, the need to pass this test drove to add the calculated dataset group in the proxy, and so, that dataset groups are reloaded too, also when opening project. But, I don't know if it is a good think if it isn't possible to remove a dataset group after calculation.

Yes, removal dataset groups is another part of the motivation, but I think this proxy could help to implement other functionalities, like changing the order of dataset groups.
But, to allow removal dataset, I am quite sure this proxy needs to be changed. More thinking is necessary.

**Is PR Finished ?**
Actualy, it seems this PR do the job (with no removal), but I don't think it is optimized, maybe ,some part need to be factored or optimized, so I add "WIP".

If you think this PR is not well designed or not do the job right way, no problem, I just wanted to purpose it, and I would move on.
If we keep it, I will try to optimize it.

**`QgsMeshDatasetIndex`**

When I look at the code, it seems this class is used to refer to dataset groups and dataset from exterior of the provider. So it it is an "external" index that it is created outside the provider and it can't store information to link external indexes with internal indexes. May be there is a solution with its implementation, but I can't see it.

vcloarec committed:
indentation corrections

PeterPetrik commented:
Hi,

I will be out until next week, so I will try to look at your changes in more detail when I return.

I think in general it is good approach to have some kind of concept of "invalid" groups so we can remove or shuffle groups. I would like to take similar approach as we have for invalid (vector) layers.  I think that XML handling and proxy handling should be part of QgsMeshLayer and not provider, since I think the behaviour should be the same for all mesh data providers. 

As I understand it now, this PR does not change any visual behaviour for user. So I think it would be good to define what we want to achieve in GUI and implement at least some of the functionality. That case we are sure that the proposed approach in the CORE part is good in the long term.

keep the good work

vcloarec commented:
Hi,

The proxy working depends on how the provider handles the dataset groups. Each mesh data provider may have a different proxy, so I don't think it could be a part of the QgsMeshLayer class. For he same reason, I think the XML handling linked to the proxy must be the responsibility of the provider through virtual methods.

The PR doesn't change a lot on visual behaviour, only that : the invalid or incompatible dataset groups still appear and they are announced as "invalid or incompatible" in the tree view of the properties dialog. For the new handling dataset functionalities in GUI : remove, add, up and down. I don"t see others...

I didn't have time to optimize the PR, tell me what do you think about it, I would adjust it when time will be available !

PeterPetrik commented:
@wonder-sk can you please add your opinion here? Thanks a lot

wonder-sk commented:
Hmm I am really confused...

@vcloarec would it be possible to explain what are the issues the PR is trying to solve? Ideally please give some use cases: eg. when I do these three steps, the expected behavior is that QGIS does this, but instead currently it does this.

I have a very basic understanding that the PR should improve handling of invalid datasets, but I do not understand why all these changes are necessary. Isn't it just possible to have a simple "invalid" flag for to dataset groups? The approach with proxies is making the code more complicated and I am probably missing something why this extra complexity is needed...

vcloarec commented:
I will try to clarify and sorry for the long text...

**When the issue appears:**
dataset groups are firtsly loaded when they are valid, and reloaded after they become invalid.
As MDAL doesn't add invalid dataset group in the mesh's dataset groups list, the indexes of datasetgroups in the list change (only for dataset groups loaded after invalid dataset groups).

**The problem:**
When QGIS calls a datasetgroup (for any reasons), QGIS uses a `QgsMeshDatasetIndex` containing a index to refer to the dataset group pointed. For the  MDAL provider, who have the responsibility to return information about the dataset groups indexed by the `QgsMeshDatasetIndex`, this index is the position of the dataset group in the list handle by MDAL. If the dataset groups is invalid, MDAL and the MDAL provider can't return a invalid flag because they don't have stored any informations about the invalid dataset group, but return the following valid dataset groups, if any.

**Example**, with this list of dataset groups firstly loaded :
                      Qgis index        MDAL index
Dataset 1              0                      0
Dataset 2              1                      1
Dataset 3              2                      2

If for any reasons, the Dataset 1 becomes invalid, the indexes are :
                      Qgis index        MDAL index
Dataset 1              0                      0
Dataset 2              1                      -
Dataset 3              2                      1

So when QGIS ask tho the MDAL provider a information about the Dataset 2 (index with value 1), the provider returns information about the Dataset 3, and when index with value of 2 is used, the provider return void objects.
Actually, the **result for the user** is that the Dataset 2 doesn't appear anymore, and the Dataset 3 is displayed with the renderer setting of the Dataset 2. It is not a big issue, especially because if the invalid dataset returns to a valid state, reloading the layer permits to return to the first state.
Maybe a more problematic issue : if the project is reopened just after a dataset group is become invalid, the invalid datasetgroups will not be reloaded and its uri will not be registered. If saving again, it can't be loaded anymore even if the dataset group returns to a valid state. The user has to add it again.  I grant there are worst cases... But it can exist cleaner behaviour.

The purposed proxy avoids this issue because its aim is to link Qgis indexes to MDAL indexes, even dataset groups are become invalid. In the purposed PR, when dataset group is invalid, instead of the behaviour described above, the provider return a void `QgsMeshDatasetGroupMetadata` with name "invalid or incompatible mesh" that appears in the view of the mesh properties dialog. It could return other information, like a invalid flag, it is not a problem. And all the valid dataset groups keep their renderer settings define by user.

This proxy principle could also open the possibility to add some functionalities like remove dataset groups or other.

Another solution is to change the MDAL behaviour, but I think the principle of providers is to be adapted to the behaviour of the source, not the opposite.

As I have said before, this is just an idea I had during discussion about reloading dataset groups, the issue is not big, so, if you think this approach is not pertinent, I will move on ...

wonder-sk commented:
Thanks a lot for the explanation, now I think I have a much better understanding of the problems you are trying to solve.

I have discussed with @PeterPetrik about how to best approach this... Apart from the reload issues you are having, there are related issues that should be solved sometime in the future:
- during a reload, one of the files creates more (or fewer) dataset groups than before (this is a generalization of your issue where reload goes from N groups to zero groups and vice versa)
- user may want to remove (hide) some dataset groups in the GUI
- user may want to reorder some dataset groups

In general neither of us is very keen to go the proposed direction with a provider specific proxy for dataset groups as it adds a fair amount of complexity and it will still fail to properly address cases where e.g. more dataset groups are loaded after reload.

The way forward in our opinion would be to stop using indexes to identify MDAL dataset groups within QGIS because indexes are never going to be stable. Instead we should start using unique identifiers (names) of dataset groups, for example for "velocity" dataset group from "/tmp/data.nc" file would have identifier (e.g. "/tmp/data.nc?name=velocity") instead of numerical index. Peter has raised a new ticket for MDAL to address that https://github.com/lutraconsulting/MDAL/issues/122 - once implemented, the new MDAL API should get used in QGIS - that will require another PR. Once this is done, all the above use cases could be handled correctly without much trouble as we would be able to always uniquely identify dataset groups. Moreover, all of that would be provider independent, which is also important.

For that reason I would suggest that we close this PR and head towards the solution described above. If the above sounds good to you, please feel free to look into contributing the necessary changes in MDAL and QGIS - we will be happy to help. (We may as well find some time to implement these changes at some point later, but it is unclear when.)

vcloarec commented:
I agree that using unique identifiers will be better than numerical indexes. If MDAL is modified in this direction it would be easier to handle link between QGIS and dataset groups.
I will be very happy to contribute to this changes but I need to find time to do this... Now, it is no my priority as I need to work on mesh editing with QGIS API, but I will try to thinking about.
So I closed this PR, hoping this PR could serve the thinking around mesh layer.
